### PR TITLE
Options refactoring

### PR DESCRIPTION
**Pull request by Slimakanzer**
This PR doesn't fix anything, but it cleans up the options module.
The options have been split into two components:
- Option Page
- Option Model

The first is responsible for presenting options on the dialog page.
The second is responsible for saving and loading options from the Visual Studio registry.
This makes the options more expandable.

This PR may also slightly speed up the extension loading.

### CODE REVIEW

File: VSRAD.Syntax/Options/GeneralOptions.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        [Description("by name")]
+        ByName = 3,
+        [Description("by name descending")]
+        ByNameDescending = 4,
+    }
+
+    public class GeneralOptions : BaseOptionModel<GeneralOptions>
+    {
+        private const string InstructionCollectionName = "RadeonAsmInstructionCollection";
+        private static readonly Regex fileExtensionRegular = new Regex(@"^\.\w+$");
```
Comment (johnny-keker) [inline VSRAD.Syntax/Options/GeneralOptions.cs:27]: Code style consistency: rename private member to `_fileExtensionRegular`?

File: VSRAD.Syntax/FunctionList/FunctionListControl.xaml.cs [hunk 2]
```diff
@@ -10,7 +10,6 @@
 using Task = System.Threading.Tasks.Task;
 using System.Threading;
 using System.Linq;
-using VSRAD.Syntax.Options;
 
 namespace VSRAD.Syntax.FunctionList
```
Comment (johnny-keker) [inline VSRAD.Syntax/FunctionList/FunctionListControl.xaml.cs:14]: In this file all private properties will be better to rename so they will start with `_`.

(for example, `private SortState SortState` is little bit confusing)

File: VSRAD.Syntax/Options/BaseOptionModel.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            ShellSettingsManager manager = await _settingsManager.GetValueAsync();
+            WritableSettingsStore settingsStore = manager.GetWritableSettingsStore(SettingsScope.UserSettings);
+
+            if (!settingsStore.CollectionExists(CollectionName))
+                settingsStore.CreateCollection(CollectionName);
+
+            foreach (PropertyInfo property in GetOptionProperties())
+            {
+                string output = SerializeValue(property.GetValue(this));
+                settingsStore.SetString(CollectionName, property.Name, output);
```
Comment (johnny-keker) [inline VSRAD.Syntax/Options/BaseOptionModel.cs:86]: Code style consistency: it is better to use `var` when possible, or always use explicit type names

File: VSRAD.Syntax/Options/BaseOptionModel.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                };
+                formatter.Serialize(stream, value);
+                stream.Flush();
+                return Convert.ToBase64String(stream.ToArray());
+            }
+        }
+
+        protected virtual object DeserializeValue(string value, Type type)
+        {
+            byte[] b = Convert.FromBase64String(value);
```
Comment (johnny-keker) [inline VSRAD.Syntax/Options/BaseOptionModel.cs:111]: Same as above

File: VSRAD.Syntax/Options/GeneralOptions.cs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    public enum SortState
+    {
+        [Description("by line number")]
+        ByLine = 1,
+        [Description("by line number descending")]
+        ByLineDescending = 2,
+        [Description("by name")]
+        ByName = 3,
+        [Description("by name descending")]
+        ByNameDescending = 4,
```
Comment (johnny-keker) [inline VSRAD.Syntax/Options/GeneralOptions.cs:21]: Explicit `int` values is unnecessary.

**Review by johnny-keker — CHANGES_REQUESTED** [review]

File: VSRAD.Syntax/Options/GeneralOptions.cs (hunk 1 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/Options/GeneralOptions.cs]: Thank you, fixed in 024da88

File: VSRAD.Syntax/Options/BaseOptionModel.cs (hunk 3 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/Options/BaseOptionModel.cs]: Thank you, fixed in 024da88

File: VSRAD.Syntax/Options/BaseOptionModel.cs (hunk 4 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/Options/BaseOptionModel.cs]: Fixed in 024da88

File: VSRAD.Syntax/Options/GeneralOptions.cs (hunk 5 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/Options/GeneralOptions.cs]: Thank you, fixed in d95146a

File: VSRAD.Syntax/FunctionList/FunctionListControl.xaml.cs (hunk 2 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/FunctionList/FunctionListControl.xaml.cs:14]: This is a good comment, but not within the scope of this pull request.

**Review by johnny-keker — APPROVED** [review]
LGTM

**Pull request merged**